CMakeLists.txt

```diff
@@ -872,6 +872,13 @@ if(SWIFT_BUILD_SWIFT_SYNTAX)
     message(WARNING "Force setting BOOTSTRAPPING=HOSTTOOLS because Swift parser integration is enabled")
     set(BOOTSTRAPPING_MODE "HOSTTOOLS")
   endif()
+
+  # We cannot mix the parser using host tools and SwiftCompilerSources uses bootstrapping
+  if(BOOTSTRAPPING_MODE MATCHES "BOOTSTRAPPING.*")
```

Comment:
Since this does exactly the same as the above block, is there a reason you didn't just add this condition there, ie changing the above check to `if(SWIFT_HOST_VARIANT_SDK MATCHES "LINUX|OPENBSD|FREEBSD" OR BOOTSTRAPPING_MODE MATCHES "BOOTSTRAPPING.*")` instead?

Reply:
Just that I wanted to separate the actual reason for them, though it's odd I didn't update the message in that case. I'll update to `if(NOT BOOTSTRAPPING_MODE STREQUAL "OFF" AND NOT BOOTSTRAPPING_MODE STREQUAL "HOSTTOOLS")` instead. Unless there was a reason we were forcing bootstrapping *on* for Linux, maybe there was 🤔

Reply:
No, that was an oversight on linux. You can just drop the second part of your new check, as it doesn't matter if it simply sets hosttools again, but are you sure you want to override the `CROSSCOMPILE*` modes also on macOS?

Reply:
[I dropped Android from that list because I didn't want `CROSSCOMPILE` overridden for Android](https://github.com/apple/swift/pull/69538/files#r1377837700).

Reply:
> You can just drop the second part of your new check, as it doesn't matter if it simply sets hosttools again

We output a warning, which I don't want if it's already hosttools.

> but are you sure you want to override the CROSSCOMPILE* modes also on macOS?

Yes, bootstrapping is being removed. First step to that is forcing hosttools, then merging SwiftCompilerSources + ASTGen, and once that's complete (or maybe at the same time), removing all the bootstrapping CMake.

> I dropped Android from that list because I didn't want CROSSCOMPILE overridden for Android

As above, bootstrapping is being removed. Would you prefer I force it to OFF here (rather than HOSTTOOLS) for Android?

Reply:
> Yes, bootstrapping is being removed. First step to that is forcing hosttools, then merging SwiftCompilerSources + ASTGen, and once that's complete (or maybe at the same time), removing all the bootstrapping CMake.

OK, but this also overrides the cross-compilation modes to build the compiler itself, eg building the compiler for macOS x86_64 on a macOS arm64 host or Android AArch64 on a linux x86_64 host. Of course, those modes are currently broken, because cross-compiling the Swift portions of the Swift compiler is currently unconfigured, but I was planning on filing an issue for that.

> As above, bootstrapping is being removed.

As above, `CROSSCOMPILE` has nothing to do with bootstrapping: it is a way to cross-compile the Swift compiler itself with an already built Swift compiler.

> Would you prefer I force it to OFF here (rather than HOSTTOOLS) for Android?

Nah, something like `if(NOT SWIFT_HOST_VARIANT_SDK STREQUAL "ANDROID" AND NOT BOOTSTRAPPING_MODE  MATCHES "HOSTTOOLS|OFF")` should be fine instead.

Reply:
To be clear, `BOOTSTRAPPING=` is being removed. Even if I do `if(NOT SWIFT_HOST_VARIANT_SDK STREQUAL "ANDROID" AND NOT BOOTSTRAPPING_MODE  MATCHES "HOSTTOOLS|OFF")`, it won't continue to work for long - the path forward is requiring a host compiler + setting the various CMake flags.

Reply:
> To be clear, BOOTSTRAPPING= is being removed.

It _should_ be renamed, but I don't think it should be _removed_, as there is still a difference between [`HOSTTOOLS`](https://github.com/apple/swift/blob/406d5337dc48af0b2159a5eccecdd8cb20bd2197/cmake/modules/AddSwift.cmake#L467) and [`CROSSCOMPILE-WITH-HOSTLIBS`](https://github.com/apple/swift/blob/406d5337dc48af0b2159a5eccecdd8cb20bd2197/cmake/modules/AddSwift.cmake#L479), that you would be overriding with this pull as currently written.

> Even if I do if(NOT SWIFT_HOST_VARIANT_SDK STREQUAL "ANDROID" AND NOT BOOTSTRAPPING_MODE  MATCHES "HOSTTOOLS|OFF"), it won't continue to work for long - the path forward is requiring a host compiler + setting the various CMake flags.

To be clear, that should be `if(NOT SWIFT_HOST_VARIANT_SDK STREQUAL "ANDROID" AND NOT BOOTSTRAPPING_MODE MATCHES "HOSTTOOLS|CROSSCOMPILE|OFF")` for now. Once you remove [the two bootstrapping modes](https://github.com/apple/swift/blob/e48b05f12c7c159a976cedd94c061f0e1f28ca52/CMakeLists.txt#L342), you can also remove this hosttools override you are modifying here altogether, but we will still need to differentiate between host builds and cross-compiles of the Swift compiler.

Reply:
> It should be renamed, but I don't think it should be removed,

`CROSSCOMPILE*` is only used to build SwiftCompilerSources, which aren't necessary to build the compiler regardless - hence my suggestion of `OFF` above. SwiftCompilerSources will end up merged with ASTGen, which doesn't (and won't) use `BOOTSTRAPPING` at all.

Reply:
> CROSSCOMPILE* is only used to build SwiftCompilerSources, which aren't necessary to build the compiler regardless - hence my suggestion of OFF above.

I haven't looked into that, but we will still need to handle cross-compiling the new swift-syntax parser too once it becomes mandatory.

> SwiftCompilerSources will end up merged with ASTGen, which doesn't (and won't) use BOOTSTRAPPING at all.

Of course, nobody wants the `BOOTSTRAPPING`* modes but there would be no reason to throw out this existing `CROSSCOMPILE`* support for SwiftCompilerSources/ASTGen then either. Rather, we should build on it to allow cross-compiling all parts of the Swift compiler that are written in Swift, instead of only the few we currently support.